### ekf 配置文件 ###
ekf_filter_node:
  ros__parameters:
# 频率（Hz），滤波器输出位置估计的速率。请注意，滤波器在收到至少一个输入消息后才会开始计算。
# 然后它将以此处指定的频率连续运行，而不管是否收到更多测量值。未指定时默认 30。
    frequency: 30.0

# 传感器超时（秒），超过该周期则认为传感器已超时。在这种情况下，我们会对 EKF 执行一次预测周期而不进行校正。
# 该参数可以被视为滤波器生成新输出的最小频率。未指定时默认为 1 / frequency。
    sensor_timeout: 0.1

# ekf_localization_node 和 ukf_localization_node 都使用 3D 全向运动模型。如果将此参数设为 true，则不会在状态估计中使用任何 3D 信息。
# 当在平面环境中运行并希望忽略例如 IMU 检测到的地面微小变化时，请使用此项。未指定时默认 false。
    two_d_mode: false
    # 频率（Hz），滤波器输出位置估计的速率。请注意，滤波器在收到至少一个输入消息后才会开始计算。
    # 然后它将以此处指定的频率连续运行，而不管是否收到更多测量值。未指定时默认 30。
    # 传感器超时（秒），超时后认为传感器失效。在这种情况下，EKF 将执行预测周期但不进行校正。
    # 该参数可被视为滤波器生成新输出的最小频率。未指定时默认为 1 / frequency。
    # ekf_localization_node 与 ukf_localization_node 使用 3D 全向运动模型。若将此参数置为 true，
    # 状态估计中将不使用任何 3D 信息。若您在平面环境中运行并希望忽略由 IMU 等检测到的地面微小变化，请使用此项。
    # 未指定时默认 false。
# 使用此参数可以为 ekf_localization_node 生成的变换提供一个偏移量。可用于将变换提前时间（future dating），
# 这对于与某些其他包交互是必需的。未指定时默认为 0.0。
    transform_time_offset: 0.0

# 使用此参数指定 tf 监听器等待变换可用的最长时间（秒）。未指定时默认为 0.0。
    transform_timeout: 0.0

# 如果遇到问题，请尝试将此项设为 true，然后 echo /diagnostics_agg 话题以查看节点是否对任何设置或数据不满意。
    print_diagnostics: true

# 调试设置。非轻量级选项。将大量信息输出到 debug_out_file 指定的文件中。注意：将其设为 true 会严重影响节点性能。未指定时默认 false。
    debug: false

# 未指定时默认为 "robot_localization_debug.txt"。请指定完整路径。
    debug_out_file: /path/to/debug/file.txt

# 是否允许旧测量触发经修正的状态重新发布
    permit_corrected_publication: false

# 是否发布加速度状态。未指定时默认 false。
    publish_acceleration: false

# 是否通过 /tf 话题广播变换。未指定时默认 true。
    publish_tf: true

# REP-105（http://www.ros.org/reps/rep-0105.html）规定了四个主要坐标框架：base_link、odom、map 和 earth。
# base_link 是附着在机器人上的坐标框架。odom 和 map 都是固定于世界的框架。
# 机器人在 odom 框架中的位置会随时间漂移，但在短期内是准确且连续的。odom 框架因此是执行局部运动规划的最佳框架。
# map 框架与 odom 框架类似，但包含更全局准确的位置信息，可能因融合 GPS 数据或基于地图的定位节点而发生离散跳变。
# earth 框架用于通过给多个 map 框架提供公共参考来关联它们。ekf_localization_node 与 ukf_localization_node 不处理 earth 框架。
# 使用以下设置方法：
# 1. 将 map_frame、odom_frame、base_link_frame 设置为系统的相应框架名称。
#     1a. 如果系统没有 map_frame，则删除它，并确保 "world_frame" 设置为 odom_frame 的值。
# 2. 如果融合的是连续位置数据（如轮速里程计、视觉里程计或 IMU 数据），将 "world_frame" 设置为 odom_frame 的值。
#    这是 robot_localization 状态估计节点的默认行为。
# 3. 如果融合的是可能发生离散跳变的全局绝对位置信息（例如 GPS 或来自地标观测的位置更新）：
#     3a. 将 "world_frame" 设置为 map_frame 的值
#     3b. 确保有其他节点生成 odom->base_link 的变换。注意这可以是另一个 robot_localization 实例，但该实例不应融合全局数据。
    map_frame: map              # 未指定时默认 "map"
    odom_frame: odom            # 未指定时默认 "odom"
    base_link_frame: base_footprint  # 未指定时默认 "base_link"
    world_frame: odom           # 未指定时默认等于 odom_frame 的值

# 该滤波器接受任意数量的每种输入消息类型（nav_msgs/Odometry、geometry_msgs/PoseWithCovarianceStamped、
# geometry_msgs/TwistWithCovarianceStamped、sensor_msgs/Imu）。要添加输入，只需在其“基名”后附加下一个序号，例如 odom0、
# odom1、twist0、twist1、imu0、imu1、imu2 等。值应为话题名称。这些参数没有默认值，必须指定。
    odom0: odom

# 每个传感器读数会更新滤波器的部分或全部状态。这些选项让您更精细地控制每个测量中哪些值被输入滤波器。
# 例如，如果有一个里程计消息作为输入，但只想使用其 Z 位置值，则将整个向量除第三项外都设为 false。
# 值的顺序为 x,      y,      z, 
#           roll,   pitch,  yaw, 
#           vx,     vy,     vz, 
#           vroll,  vpitch, vyaw, 
#           ax,     ay,     az。
# 注意某些消息类型并不提供滤波器估计的所有状态变量。例如，TwistWithCovarianceStamped 没有位姿信息，
# 因此前六个值在该情况下没有意义。每个向量如果未指定则默认全为 false，因此对每个传感器此参数实际上是必需的。
    odom0_config: [false,  false,  false,
                  false, false, false,
                  true, false, false,
                  false, false, false,
                  false, false, false]
    # odom0_config: [true,  true,  false,
    #                 false, false, false,
    #                 true,  false, false,
    #                 false, false, false,
    #                 false, false, false]

# 如果有高频数据或运行频率较低，可能希望增大订阅队列大小以便融合更多测量。
    odom0_queue_size: 2

# [高级] ROS 中的大消息在高频到达时可能会表现异常，这是由于 Nagle 算法造成的。此选项指示 ROS 订阅器使用 tcpNoDelay，
# 即禁用 Nagle 算法。
    odom0_nodelay: false

# [高级] 当用两个传感器测量同一位姿变量时，可能出现二者都低估协方差的情况。这会导致滤波器在两者到达时来回跳变。
# 在这些情况下，通常的做法是 (a) 修正测量协方差，或 (b) 如果其中一个传感器也测量速度，则让一个传感器测量位姿，另一个测量速度。
# 如果无法做到 (a) 或 (b)，我们公开了 differential 参数。当启用差分模式时，所有绝对位姿数据会通过对绝对位姿测量求差分转换为速度数据。
# 这些速度随后像往常一样被积分。注意：这仅适用于提供位姿测量的传感器；对 twist 测量设置 differential 为 true 无效。
    odom0_differential: false

# [高级] 节点启动时，如果此参数为 true，则第一个测量被视为所有未来测量的“零点”。虽然可以通过 differential 参数实现类似效果，
# 关键区别是 relative 不会将测量转换为速度再积分。如果只是想让给定传感器的测量从 0 开始，请将此项设为 true。
    odom0_relative: false

# [高级] 如果数据可能包含离群点，使用这些以马氏距离表示的阈值来控制传感器测量允许距离当前车辆状态的最大偏差。
# 每个阈值若未指定则默认为 numeric_limits<double>::max()。强烈建议在不需要时移除这些参数。
# 对于同时包含位姿和速度数据的消息，此参数指定阈值应用于消息的哪一部分（位姿或速度）。
    odom0_pose_rejection_threshold: 5.0
    odom0_twist_rejection_threshold: 1.0


    imu0: imu
    imu0_config: [false, false, false,
            false, false,  true,
            false, false, false,
            false, false,  true,
            false, false,  false]
    imu0_nodelay: false
    imu0_differential: false
    imu0_relative: true
    imu0_queue_size: 5
    imu0_pose_rejection_threshold: 0.8                 # 注意参数名的差别
    imu0_twist_rejection_threshold: 0.8                #
    imu0_linear_acceleration_rejection_threshold: 0.8  #

# [高级] 一些 IMU 会自动去除重力加速度，而另一些不会。如果您的 IMU 没有去除，请将此项设为 true，且*确保*您的数据符合 REP-103，
# 特别是数据采用 ENU 坐标系。
    imu0_remove_gravitational_acceleration: true

# [高级] EKF 和 UKF 模型遵循标准的 predict/correct 周期。在预测阶段，如果没有加速度参考，t+1 时刻的速度被预测为与 t 时刻相同。
# 在校正阶段，这一预测值会与测量值融合以生成新的速度估计。这可能导致最终速度在旧速度和新测量之间成为加权平均，从而使位姿积分收敛缓慢。
# 在旋转过程中使用激光雷达数据时，这种现象尤其明显。为了解决该问题，用户可以尝试增大速度变量的 process_noise_covariance 对角元素，
# 或减小测量中的该变量方差。此外，用户也可以利用控制命令在预测时提供的加速度参考。如果使用控制输入，它会被转换为加速度项并用于预测。
# 注意，如果某个变量在输入中已有加速度测量，则会忽略控制项。
# 是否在预测阶段使用控制输入。未指定时默认 false。
    use_control: true
# 输入（假定为 cmd_vel）是 geometry_msgs/Twist 还是 geometry_msgs/TwistStamped。未指定时默认 false（即 Twist）。
    stamped_control: false
# 最后发出的控制命令将在此时长内用于预测。未指定时默认 0.2。
    control_timeout: 0.2
# 哪些速度受到控制。顺序为 vx, vy, vz, vroll, vpitch, vyaw。
    control_config: [true, false, false, false, false, true]
# 对加速度项的大小进行限制。应与机器人运动学匹配。
    acceleration_limits: [1.3, 0.0, 0.0, 0.0, 0.0, 3.4]
# 加速与减速限制不一定相同。
    deceleration_limits: [1.3, 0.0, 0.0, 0.0, 0.0, 4.5]
# 如果机器人无法瞬时达到其加速度极限，可通过这些增益控制允许的变化
    acceleration_gains: [0.8, 0.0, 0.0, 0.0, 0.0, 0.9]
# 如果机器人无法瞬时达到其减速度极限，可通过这些增益控制允许的变化
    deceleration_gains: [1.0, 0.0, 0.0, 0.0, 0.0, 1.0]
# [高级] 过程噪声协方差矩阵难以调优，并且可能因应用而异，因此作为配置参数公开。
# 该矩阵表示在每次预测步骤后我们添加到总体误差中的噪声。全向运动模型越符合系统，这些值可以越小。
# 然而，如果某个变量收敛缓慢，可以增加该变量对应的 process_noise_covariance 对角值，
# 这会导致滤波器的预测误差增大，从而在校正时更信任传入测量。值的顺序为 x, y, z, roll, pitch, yaw,
# vx, vy, vz, vroll, vpitch, vyaw, ax, ay, az。未指定时默认为下列矩阵。
    process_noise_covariance: [0.05, 0.0,    0.0,    0.0,    0.0,    0.0,    0.0,     0.0,     0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,
                   0.0,    0.05, 0.0,    0.0,    0.0,    0.0,    0.0,     0.0,     0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,
                   0.0,    0.0,    0.06, 0.0,    0.0,    0.0,    0.0,     0.0,     0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,
                   0.0,    0.0,    0.0,    0.03, 0.0,    0.0,    0.0,     0.0,     0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,
                   0.0,    0.0,    0.0,    0.0,    0.03, 0.0,    0.0,     0.0,     0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,
                   0.0,    0.0,    0.0,    0.0,    0.0,    0.06, 0.0,     0.0,     0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,
                   0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.025, 0.0,     0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,
                   0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,     0.025, 0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,
                   0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,     0.0,     0.04, 0.0,    0.0,    0.0,    0.0,    0.0,    0.0,
                   0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,     0.0,     0.0,    0.01, 0.0,    0.0,    0.0,    0.0,    0.0,
                   0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,     0.0,     0.0,    0.0,    0.01, 0.0,    0.0,    0.0,    0.0,
                   0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,     0.0,     0.0,    0.0,    0.0,    0.02, 0.0,    0.0,    0.0,
                   0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,     0.0,     0.0,    0.0,    0.0,    0.0,    0.01, 0.0,    0.0,
                   0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,     0.0,     0.0,    0.0,    0.0,    0.0,    0.0,    0.01, 0.0,
                   0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,     0.0,     0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.015]
# [高级] 这表示状态估计误差协方差矩阵（初始值）。将某个对角值（方差）设为较大值会导致该变量对初始测量的快速收敛。
# 用户应注意不要对不会被直接测量的变量使用过大的值。值的顺序为 x, y, z, roll, pitch, yaw, vx, vy, vz,
# vroll, vpitch, vyaw, ax, ay, az。未指定时默认为下列矩阵。
    initial_estimate_covariance: [1e-9, 0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,     0.0,     0.0,     0.0,    0.0,    0.0,
                    0.0,    1e-9, 0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,     0.0,     0.0,     0.0,    0.0,    0.0,
                    0.0,    0.0,    1e-9, 0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,     0.0,     0.0,     0.0,    0.0,    0.0,
                    0.0,    0.0,    0.0,    1e-9, 0.0,    0.0,    0.0,    0.0,    0.0,    0.0,     0.0,     0.0,     0.0,    0.0,    0.0,
                    0.0,    0.0,    0.0,    0.0,    1e-9, 0.0,    0.0,    0.0,    0.0,    0.0,     0.0,     0.0,     0.0,    0.0,    0.0,
                    0.0,    0.0,    0.0,    0.0,    0.0,    1e-9, 0.0,    0.0,    0.0,    0.0,     0.0,     0.0,     0.0,    0.0,    0.0,
                    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    1e-9, 0.0,    0.0,    0.0,     0.0,     0.0,     0.0,    0.0,    0.0,
                    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    1e-9, 0.0,    0.0,     0.0,     0.0,     0.0,    0.0,    0.0,
                    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    1e-9, 0.0,     0.0,     0.0,     0.0,    0.0,    0.0,
                    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    1e-9,  0.0,     0.0,     0.0,    0.0,    0.0,
                    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,     1e-9,  0.0,     0.0,    0.0,    0.0,
                    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,     0.0,     1e-9,  0.0,    0.0,    0.0,
                    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,     0.0,     0.0,     1e-9, 0.0,    0.0,
                    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,     0.0,     0.0,     0.0,    1e-9, 0.0,
                    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,     0.0,     0.0,     0.0,    0.0,    1e-9]
